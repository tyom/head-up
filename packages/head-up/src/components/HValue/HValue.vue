<template>
  <dl class="HValue">
    <dt class="label">{{ label }}</dt>
    <dd class="value">
      <v-icon
        v-if="hasIcon"
        :name="increase ? 'caret-up' : 'caret-down'"
        :class="increase ? 'change-increase' : 'change-decrease'"
      />
      <template v-if="value">
        {{ tween && tweenedValue || value | format(formatNumber) }}
      </template>
      <template v-else>
        N/A
      </template>
    </dd>
  </dl>
</template>

<script>
import { tween } from 'femtotween';

export default {
  name: 'HValue',
  props: {
    label: {
      type: String,
      default: '',
    },
    formatNumber: {
      type: Object,
      default: () => {},
    },
    value: {
      type: [String, Number],
      default: '',
    },
    increase: {
      type: Boolean,
      default: false,
    },
    decrease: {
      type: Boolean,
      default: false,
    },
    tween: {
      type: Boolean,
      default: true,
    },
    tweenFixed: {
      type: Number,
      default: 3,
    },
  },
  computed: {
    hasIcon() {
      return Boolean(this.increase) || Boolean(this.decrease);
    },
  },
  data() {
    return {
      tweenedValue: this.value,
    };
  },
  watch: {
    value(newVal, oldVal) {
      tween(oldVal, newVal, v => {
        const newNumber = Number(v);
        this.tweenedValue = newNumber ? newNumber.toFixed(this.tweenFixed) : v;
      });
    },
  },
  filters: {
    format(value, options) {
      return Number(value).toLocaleString(undefined, options);
    },
  },
};
</script>

<style scoped>
.HValue {
  text-align: center;
}

.value {
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 1.4em;
}

.fa-icon {
  width: 1.2em;
  height: 1.2em;
  margin-right: 0.2em;
}

.change-increase {
  color: #00c200;
}

.change-decrease {
  color: #ff313e;
}
</style>
