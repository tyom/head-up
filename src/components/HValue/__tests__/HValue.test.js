import { shallowMount } from '@vue/test-utils';
import { tween } from 'femtotween';
import HValue from '../HValue';

jest.mock('femtotween', () => ({
  tween: jest.fn((oldVal, newVal, cb) => cb(newVal)),
}));

test('render default', () => {
  const wrapper = shallowMount(HValue);
  expect(wrapper).toMatchSnapshot();
});

test('render with label and value', () => {
  const wrapper = shallowMount(HValue, {
    propsData: {
      label: 'Strength',
      value: 100,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render with value from scoped slot', () => {
  const wrapper = shallowMount(HValue, {
    propsData: {
      label: 'Strength',
      value: 100,
    },
    scopedSlots: {
      value: '<h1 slot-scope="{value}">{{ value }}%</h1>',
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render with pre- and post-value slots', () => {
  const wrapper = shallowMount(HValue, {
    propsData: {
      label: 'Strength',
      value: 100,
    },
    slots: {
      'pre-value': '<h1>before value</h1>',
      'post-value': '<h1>after value</h1>',
    },
  });
  expect(wrapper).toMatchSnapshot();
});

describe('tweened value', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(HValue, {
      propsData: {
        label: 'Strength',
        value: 100,
      },
    });
  });

  test('calls tween function', async () => {
    wrapper.setProps({
      value: 120,
    });

    await wrapper.vm.$nextTick();
    expect(tween).toHaveBeenCalledWith(100, 120, expect.any(Function));
  });

  test('sets tweenedValue data property', async () => {
    wrapper.setProps({
      value: 123.456789,
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.vm.tweenedValue).toEqual('123.46');

    wrapper.setProps({
      value: 'purple',
    });

    await wrapper.vm.$nextTick();
    expect(wrapper.vm.tweenedValue).toEqual('purple');
  });
});
