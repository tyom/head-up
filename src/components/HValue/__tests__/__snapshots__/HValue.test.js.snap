// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render default 1`] = `
<dl class="HValue">
  <dt class="label"></dt>
  <dd class="value-wrapper"> <span class="value _na">
      N/A
    </span> </dd>
</dl>
`;

exports[`render with label and value 1`] = `
<dl class="HValue">
  <dt class="label">Strength</dt>
  <dd class="value-wrapper"> <span class="value">100</span> </dd>
</dl>
`;

exports[`render with pre- and post-value slots 1`] = `
<dl class="HValue">
  <dt class="label">Strength</dt>
  <dd class="value-wrapper">
    <h1>before value</h1> <span class="value">100</span>
    <h1>after value</h1>
  </dd>
</dl>
`;

exports[`render with value from scoped slot 1`] = `
<dl class="HValue">
  <dt class="label">Strength</dt>
  <dd class="value-wrapper"> <span class="value"><h1>100%</h1></span> </dd>
</dl>
`;
